diff -Naur mednafen-pure/src/pcfx/mem-handler.inc mednafen/src/pcfx/mem-handler.inc
--- mednafen-pure/src/pcfx/mem-handler.inc	2019-01-27 22:52:37.000000000 -0800
+++ mednafen/src/pcfx/mem-handler.inc	2019-03-07 17:31:20.682868200 -0800
@@ -21,7 +21,7 @@
 
 uint8 MDFN_FASTCALL mem_peekbyte(const v810_timestamp_t timestamp, const uint32 A)
 {
- if(A <= 0x001FFFFF)
+ if(A <= 0x007FFFFF)
   return(RAM[A]);
  else if(A <= 0x00FFFFFF)
   return(0xFF);
@@ -36,7 +36,7 @@
 
 uint16 MDFN_FASTCALL mem_peekhword(const v810_timestamp_t timestamp, const uint32 A) // TODO: Full memory map peeking.
 {
- if(A <= 0x001FFFFF)
+ if(A <= 0x007FFFFF)
   return MDFN_de16lsb<true>(&RAM[A]);
  else if(A <= 0x00FFFFFF)
   return 0xFFFF;
@@ -50,7 +50,7 @@
 
 static uint8 MDFN_FASTCALL mem_rbyte(v810_timestamp_t &timestamp, uint32 A)
 {
- if(A <= 0x001FFFFF)
+ if(A <= 0x007FFFFF)
  {
   RAMLPCHECK;
   return(RAM[A]);
@@ -97,7 +97,7 @@
 
 static uint16 MDFN_FASTCALL mem_rhword(v810_timestamp_t &timestamp, uint32 A)
 {
- if(A <= 0x001FFFFF)
+ if(A <= 0x007FFFFF)
  {
   RAMLPCHECK;
   return MDFN_de16lsb<true>(&RAM[A]);
@@ -176,7 +176,7 @@
 
 static uint32 MDFN_FASTCALL mem_rword(v810_timestamp_t &timestamp, uint32 A)
 {
- if(A <= 0x001FFFFF)
+ if(A <= 0x007FFFFF)
  {
   RAMLPCHECK;
   return MDFN_de32lsb<true>(&RAM[A]);
@@ -209,7 +209,7 @@
 
 static void MDFN_FASTCALL mem_wbyte(v810_timestamp_t &timestamp, uint32 A, uint8 V)
 {
- if(A <= 0x001FFFFF)
+ if(A <= 0x007FFFFF)
  {
   RAMLPCHECK;
   RAM[A] = V;
@@ -256,7 +256,7 @@
 
 static void MDFN_FASTCALL mem_whword(v810_timestamp_t &timestamp, uint32 A, uint16 V)
 {
- if(A <= 0x001FFFFF)
+ if(A <= 0x007FFFFF)
  {
   RAMLPCHECK;
   MDFN_en16lsb<true>(&RAM[A], V);
@@ -332,7 +332,7 @@
 
 static void MDFN_FASTCALL mem_wword(v810_timestamp_t &timestamp, uint32 A, uint32 V)
 {
- if(A <= 0x001FFFFF)
+ if(A <= 0x007FFFFF)
  {
   RAMLPCHECK;
   MDFN_en32lsb<true>(&RAM[A], V);
diff -Naur mednafen-pure/src/pcfx/pcfx.cpp mednafen/src/pcfx/pcfx.cpp
--- mednafen-pure/src/pcfx/pcfx.cpp	2019-01-27 22:52:37.000000000 -0800
+++ mednafen/src/pcfx/pcfx.cpp	2019-03-07 17:31:20.698468200 -0800
@@ -338,7 +338,7 @@
  Last_VDC_AR[0] = 0;
  Last_VDC_AR[1] = 0;
 
- memset(RAM, 0x00, 2048 * 1024);
+ memset(RAM, 0x00, 8192 * 1024);
 
  for(int i = 0; i < 2; i++)
  {
@@ -390,7 +390,7 @@
  {
   while(Length--)
   {
-   Address &= 2048 * 1024 - 1;
+   Address &= 8192 * 1024 - 1;
    *Buffer = RAM[Address];
    Address++;
    Buffer++;
@@ -464,9 +464,9 @@
    {
     BIOSROM[Address & 0xFFFFF] = *Buffer;
    }
-   else if(Address <= 0x1FFFFF)
+   else if(Address <= 0x7FFFFF)
    {
-    RAM[Address & 0x1FFFFF] = *Buffer;
+    RAM[Address & 0x7FFFFF] = *Buffer;
    }
    else if(Address >= 0xE0000000 && Address <= 0xE7FFFFFF)
    {
@@ -492,7 +492,7 @@
  {
   while(Length--)
   {
-   Address &= 2048 * 1024 - 1;
+   Address &= 8192 * 1024 - 1;
    RAM[Address] = *Buffer;
    Address++;
    Buffer++;
@@ -591,7 +591,7 @@
  uint32 RAM_Map_Addresses[1] = { 0x00000000 };
  uint32 BIOSROM_Map_Addresses[1] = { 0xFFF00000 };
 
- RAM = PCFX_V810.SetFastMap(RAM_Map_Addresses, 0x00200000, 1, _("RAM"));
+ RAM = PCFX_V810.SetFastMap(RAM_Map_Addresses, 0x00800000, 1, _("RAM"));
  BIOSROM = PCFX_V810.SetFastMap(BIOSROM_Map_Addresses, 0x00100000, 1, _("BIOS ROM"));
 
  {
@@ -625,7 +625,7 @@
 
  #ifdef WANT_DEBUGGER
  ASpace_Add(PCFXDBG_GetAddressSpaceBytes, PCFXDBG_PutAddressSpaceBytes, "cpu", "CPU Physical", 32);
- ASpace_Add(PCFXDBG_GetAddressSpaceBytes, PCFXDBG_PutAddressSpaceBytes, "ram", "RAM", 21);
+ ASpace_Add(PCFXDBG_GetAddressSpaceBytes, PCFXDBG_PutAddressSpaceBytes, "ram", "RAM", 23);
  ASpace_Add(PCFXDBG_GetAddressSpaceBytes, PCFXDBG_PutAddressSpaceBytes, "backup", "Internal Backup Memory", 15);
  ASpace_Add(PCFXDBG_GetAddressSpaceBytes, PCFXDBG_PutAddressSpaceBytes, "exbackup", "External Backup Memory", 15);
  ASpace_Add(PCFXDBG_GetAddressSpaceBytes, PCFXDBG_PutAddressSpaceBytes, "bios", "BIOS ROM", 20);
@@ -691,7 +691,7 @@
  MDFNGameInfo->lcm_height = MDFNGameInfo->nominal_height;
 
  MDFNMP_Init(1024 * 1024, ((uint64)1 << 32) / (1024 * 1024));
- MDFNMP_AddRAM(2048 * 1024, 0x00000000, RAM);
+ MDFNMP_AddRAM(8192 * 1024, 0x00000000, RAM);
 
 
  BackupSignalDirty = false;
@@ -985,7 +985,7 @@
 
  SFORMAT StateRegs[] =
  {
-  SFPTR8(RAM, 0x200000),
+  SFPTR8(RAM, 0x800000),
   SFVAR(Last_VDC_AR),
   SFVAR(RAM_LPA),
   SFVAR(BackupControl),
